exercises/diamond/test/Tests.hs

```diff
@@ -107,3 +142,57 @@ cases = [ Case { description = "Degenerate case with a single 'A' row"
                         "                         A                         "]
                }
         ]
+
+genNonAlphaChar :: Gen Char
+genNonAlphaChar = arbitraryASCIIChar `suchThat` (not . isLetter)
+
+genAlphaChar :: Gen Char
+genAlphaChar = elements ['A'..'Z']
+
+genDiamond :: Gen (Char, Maybe [String])
+genDiamond = do
+  c <- genAlphaChar
+  return (c, diamond c)
+
+forAllDiamond :: Testable prop => ([String] -> prop) -> Property
+forAllDiamond = forAllCharDiamond . const
+
+forAllCharDiamond :: Testable prop => (Char -> [String] -> prop) -> Property
+forAllCharDiamond p = forAll genDiamond $ uncurry p . fmap (fromMaybe [""])
+
+topLength :: [String] -> Int
+topLength = (`div` 2) . length
+
+position :: Char -> Int
+position c = ord c - ord 'A'
+
+topEqualToBottom :: [String] -> Bool
+topEqualToBottom xs = take len xs == take len (reverse xs)
+  where
+    len = topLength xs
+
+checkMiddle :: Char -> [String] -> Bool
+checkMiddle c xs = getMiddle xs == middle
+  where
+    getMiddle ys = ys !! topLength ys
+    leftSide = c : replicate (position c) ' '
+    rightSide = tail . reverse $ leftSide
+    middle = leftSide ++ rightSide
```

Comment:
Along with all the other predicates at the bottom of the file, I will recommend that `checkMiddle` is also moved up.

I will also say the way this property is expressed is very close to being a re-implementation.

An exaggerated example of what I mean is:

Imagine that you property-test `isLeapYear` in the following way:

```haskell
newtype Year = Year Int deriving (Eq, Show)
instance Arbitrary Year where
  arbitrary = Year <$> (genYear :: Gen Int)

prop_isLeapYear :: Year -> Bool
prop_isLeapYear (Year year)
  | year `rem` 400 = isLeapYear year
  | year `rem` 100 = not (isLeapYear year)
  | year `rem`   4 = isLeapYear year
  | otherwise      = False 
```

The property then becomes "the student solution is equivalent to an embedded reference solution". Having a competing implementation that you can check against is indeed a very effective way to verify that your implementation is correct. But embedding it in a test suite for an exercise is not so ideal.

Having re-read [Mark Seemann's blog post](https://blog.ploeh.dk/2015/01/10/diamond-kata-with-fscheck/), I can't really find any properties that he tests that are equivalent to checking the middle part, so perhaps it is fine to skip on this.

Reply:
I am fine removing this, see f7b40857f6a8d11b081efca432b4cfb405201095